README.md
An isomorphic javascript library with typescript type definitions that provides a mechanism to access Azure Endpoints in different Azure clouds. It also provides a mechanism to add a custom environment.

### Example

```javascript
import { Environment, EnvironmentParameters } from "ms-rest-azure-env";

// Accesing predefined environment endpoints.
console.log(Environment.AzureCloud.resourceManagerEndpointUrl);
console.log(Environment.ChinaCloud.resourceManagerEndpointUrl);

//Adding a custom environment with required endpoint values.
let df: AzureEnvironmentParameters = {
  name: "Dogfood",
  portalUrl: "http://go.microsoft.com/fwlink/?LinkId=254433",
  managementEndpointUrl: "https://management.core.windows.net",
  resourceManagerEndpointUrl: "https://management.azure.com/",
  activeDirectoryEndpointUrl: "https://login.microsoftonline.com/",
  activeDirectoryResourceId: "https://management.core.windows.net/"
};
Environment.add(df);
let dfoodEnv = Environment.get("Dogfood");
console.log(dfoodEnv);
console.log(`Accessing the custom environment info: ${(<any>Environment)["Dogfood"].managementEndpointUrl}`);
```

### Contributing

This project welcomes contributions and suggestions.  Most contributions require you to agree to a
Contributor License Agreement (CLA) declaring that you have the right to, and actually do, grant us
the rights to use your contribution. For details, visit https://cla.microsoft.com.

When you submit a pull request, a CLA-bot will automatically determine whether you need to provide
a CLA and decorate the PR appropriately (e.g., label, comment). Simply follow the instructions
provided by the bot. You will only need to do this once across all repos using our CLA.

This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/).
For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or
contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.


Changelog.md
### 1.0.0 - 2018/10/04

- Use rollup for bundling
- Use global name Azure.Environment instead of AzureEnvironment class

### 0.1.2 - 2018/08/28

- Update ActiveDirectory endpoint URL in US Government cloud

### 0.1.1 - 2018/08/24

- added batchResourceId to the supported environments

### 0.1.0 - 2017/09/16

- Initial version of ms-rest-azure-env
  - An isomorphic library that provides information about Azure endpoints for known Azure Clouds.
  - Provides a mechanism to add custom environments.


test/azureEnvironmentTests.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import * as should from "should";
import { Environment, EnvironmentParameters } from "../lib/azureEnvironment";

describe("AzureEnvironment", function () {

  it("should show the details of Azure Production environment correctly", function (done) {
    const tempEnv = Environment.AzureCloud;
    tempEnv.name.should.equal("AzureCloud");
    tempEnv.batchResourceId.should.equal("https://batch.core.windows.net/");
    tempEnv.activeDirectoryEndpointUrl.should.equal("https://login.microsoftonline.com/");
    tempEnv.activeDirectoryResourceId.should.equal("https://management.core.windows.net/");
    tempEnv.managementEndpointUrl.should.equal("https://management.core.windows.net");
    tempEnv.resourceManagerEndpointUrl.should.equal("https://management.azure.com/");
    tempEnv.portalUrl.should.equal("https://portal.azure.com");
    should.equal(tempEnv.validateAuthority, true);
    done();
  });

  it("should show Environment.ChinaCloud environment correctly", function (done) {
    const tempEnv = Environment.ChinaCloud;
    tempEnv.name.should.equal("AzureChinaCloud");
    tempEnv.batchResourceId.should.equal("https://batch.chinacloudapi.cn/");
    tempEnv.activeDirectoryEndpointUrl.should.equal("https://login.chinacloudapi.cn/");
    tempEnv.activeDirectoryResourceId.should.equal("https://management.core.chinacloudapi.cn/");
    tempEnv.managementEndpointUrl.should.equal("https://management.core.chinacloudapi.cn");
    tempEnv.resourceManagerEndpointUrl.should.equal("https://management.chinacloudapi.cn");
    tempEnv.portalUrl.should.equal("https://portal.azure.cn");
    tempEnv.validateAuthority.should.equal(true);
    done();
  });

  it("should show the details of Azure USGovernment environment correctly", function (done) {
    const tempEnv = Environment.USGovernment;
    tempEnv.name.should.equal("AzureUSGovernment");
    tempEnv.batchResourceId.should.equal("https://batch.core.usgovcloudapi.net/");
    tempEnv.activeDirectoryEndpointUrl.should.equal("https://login.microsoftonline.us/");
    tempEnv.activeDirectoryResourceId.should.equal("https://management.core.usgovcloudapi.net/");
    tempEnv.managementEndpointUrl.should.equal("https://management.core.usgovcloudapi.net");
    tempEnv.resourceManagerEndpointUrl.should.equal("https://management.usgovcloudapi.net");
    tempEnv.portalUrl.should.equal("https://portal.azure.us");
    tempEnv.validateAuthority.should.equal(true);
    done();
  });

  it("should show the details of Azure GermanCloud environment correctly", function (done) {
    const tempEnv = Environment.GermanCloud;
    tempEnv.name.should.equal("AzureGermanCloud");
    tempEnv.batchResourceId.should.equal("https://batch.microsoftazure.de/");
    tempEnv.activeDirectoryEndpointUrl.should.equal("https://login.microsoftonline.de/");
    tempEnv.activeDirectoryResourceId.should.equal("https://management.core.cloudapi.de/");
    tempEnv.managementEndpointUrl.should.equal("https://management.core.cloudapi.de");
    tempEnv.resourceManagerEndpointUrl.should.equal("https://management.microsoftazure.de");
    tempEnv.portalUrl.should.equal("https://portal.microsoftazure.de/");
    tempEnv.validateAuthority.should.equal(true);
    done();
  });

  it("should be able to add a new environment", function (done) {
    const df: EnvironmentParameters = {
      name: "Dogfood",
      portalUrl: "http://go.microsoft.com/fwlink/?LinkId=254433",
      managementEndpointUrl: "https://management.core.windows.net",
      resourceManagerEndpointUrl: "https://management.azure.com/",
      activeDirectoryEndpointUrl: "https://login.microsoftonline.com/",
      activeDirectoryResourceId: "https://management.core.windows.net/"
    };
    const tempEnv = Environment;
    tempEnv.add(df);
    const dfood = tempEnv.get(df.name);
    dfood.name.should.equal("Dogfood");
    dfood.activeDirectoryEndpointUrl.should.equal("https://login.microsoftonline.com/");
    dfood.activeDirectoryResourceId.should.equal("https://management.core.windows.net/");
    dfood.managementEndpointUrl.should.equal("https://management.core.windows.net");
    dfood.resourceManagerEndpointUrl.should.equal("https://management.azure.com/");
    dfood.portalUrl.should.equal("http://go.microsoft.com/fwlink/?LinkId=254433");
    should.equal(dfood.validateAuthority, true);

    // Verify that the environment properly got added to the prototype
    (<any>tempEnv)["Dogfood"].name.should.equal("Dogfood");
    (<any>tempEnv)["Dogfood"].activeDirectoryEndpointUrl.should.equal("https://login.microsoftonline.com/");
    (<any>tempEnv)["Dogfood"].activeDirectoryResourceId.should.equal("https://management.core.windows.net/");
    (<any>tempEnv)["Dogfood"].managementEndpointUrl.should.equal("https://management.core.windows.net");
    (<any>tempEnv)["Dogfood"].resourceManagerEndpointUrl.should.equal("https://management.azure.com/");
    (<any>tempEnv)["Dogfood"].portalUrl.should.equal("http://go.microsoft.com/fwlink/?LinkId=254433");
    (<any>tempEnv)["Dogfood"].validateAuthority.should.equal(true);
    done();
  });
});

azureEnvironmentBundle.js
var azureEnvironment=function(e){var t={};function r(a){if(t[a])return t[a].exports;var n=t[a]={i:a,l:!1,exports:{}};return e[a].call(n.exports,n,n.exports,r),n.l=!0,n.exports}return r.m=e,r.c=t,r.d=function(e,t,a){r.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:a})},r.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},r.t=function(e,t){if(1&t&&(e=r(e)),8&t)return e;if(4&t&&"object"==typeof e&&e&&e.__esModule)return e;var a=Object.create(null);if(r.r(a),Object.defineProperty(a,"default",{enumerable:!0,value:e}),2&t&&"string"!=typeof e)for(var n in e)r.d(a,n,function(t){return e[t]}.bind(null,n));return a},r.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return r.d(t,"a",t),t},r.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},r.p="",r(r.s=0)}([function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0});class a{constructor(e){if(this.validateAuthority=!0,!e||"object"!=typeof e)throw new Error("'parameters' is a required parameter and must be of type 'object'.");["name","portalUrl","managementEndpointUrl","resourceManagerEndpointUrl","activeDirectoryEndpointUrl","activeDirectoryResourceId"].forEach(function(t){if(!e[t]||"string"!=typeof e[t].valueOf())throw new Error(`Please provide "${t}" for the environment and it must be of type "string".`)}),this.name=e.name,this.portalUrl=e.portalUrl,this.managementEndpointUrl=e.managementEndpointUrl,this.resourceManagerEndpointUrl=e.resourceManagerEndpointUrl,this.activeDirectoryEndpointUrl=e.activeDirectoryEndpointUrl,this.activeDirectoryResourceId=e.activeDirectoryResourceId,this.activeDirectoryGraphApiVersion&&(this.activeDirectoryGraphApiVersion=e.activeDirectoryGraphApiVersion),this.activeDirectoryGraphResourceId&&(this.activeDirectoryGraphResourceId=e.activeDirectoryGraphResourceId),this.azureDataLakeAnalyticsCatalogAndJobEndpointSuffix&&(this.azureDataLakeAnalyticsCatalogAndJobEndpointSuffix=e.azureDataLakeAnalyticsCatalogAndJobEndpointSuffix),this.azureDataLakeStoreFileSystemEndpointSuffix&&(this.azureDataLakeStoreFileSystemEndpointSuffix=e.azureDataLakeStoreFileSystemEndpointSuffix),this.batchResourceId&&(this.batchResourceId=e.batchResourceId),this.galleryEndpointUrl&&(this.galleryEndpointUrl=e.galleryEndpointUrl),this.keyVaultDnsSuffix&&(this.keyVaultDnsSuffix=e.keyVaultDnsSuffix),this.publishingProfileUrl&&(this.publishingProfileUrl=e.publishingProfileUrl),this.sqlManagementEndpointUrl&&(this.sqlManagementEndpointUrl=e.sqlManagementEndpointUrl),this.sqlServerHostnameSuffix&&(this.sqlServerHostnameSuffix=e.sqlServerHostnameSuffix),this.storageEndpointSuffix&&(this.storageEndpointSuffix=e.storageEndpointSuffix)}static add(e){const t={},r=new a(e);t[e.name]=r,Object.assign(a,t)}static get(e){if(!e)throw new TypeError("name cannot be null or undefined and must be of type string.");return a[e]}}a.Azure={name:"Azure",portalUrl:"https://portal.azure.com",publishingProfileUrl:"https://go.microsoft.com/fwlink/?LinkId=254432",managementEndpointUrl:"https://management.core.windows.net",resourceManagerEndpointUrl:"https://management.azure.com/",sqlManagementEndpointUrl:"https://management.core.windows.net:8443/",sqlServerHostnameSuffix:".database.windows.net",galleryEndpointUrl:"https://gallery.azure.com/",activeDirectoryEndpointUrl:"https://login.microsoftonline.com/",activeDirectoryResourceId:"https://management.core.windows.net/",activeDirectoryGraphResourceId:"https://graph.windows.net/",batchResourceId:"https://batch.core.windows.net/",activeDirectoryGraphApiVersion:"2013-04-05",storageEndpointSuffix:".core.windows.net",keyVaultDnsSuffix:".vault.azure.net",azureDataLakeStoreFileSystemEndpointSuffix:"azuredatalakestore.net",azureDataLakeAnalyticsCatalogAndJobEndpointSuffix:"azuredatalakeanalytics.net",validateAuthority:!0},a.AzureChina={name:"AzureChina",portalUrl:"https://portal.azure.cn",publishingProfileUrl:"https://go.microsoft.com/fwlink/?LinkID=301774",managementEndpointUrl:"https://management.core.chinacloudapi.cn",resourceManagerEndpointUrl:"https://management.chinacloudapi.cn",sqlManagementEndpointUrl:"https://management.core.chinacloudapi.cn:8443/",sqlServerHostnameSuffix:".database.chinacloudapi.cn",galleryEndpointUrl:"https://gallery.chinacloudapi.cn/",activeDirectoryEndpointUrl:"https://login.chinacloudapi.cn/",activeDirectoryResourceId:"https://management.core.chinacloudapi.cn/",activeDirectoryGraphResourceId:"https://graph.chinacloudapi.cn/",activeDirectoryGraphApiVersion:"2013-04-05",batchResourceId:"https://batch.chinacloudapi.cn/",storageEndpointSuffix:".core.chinacloudapi.cn",keyVaultDnsSuffix:".vault.azure.cn",azureDataLakeStoreFileSystemEndpointSuffix:"N/A",azureDataLakeAnalyticsCatalogAndJobEndpointSuffix:"N/A",validateAuthority:!0},a.AzureUSGovernment={name:"AzureUSGovernment",portalUrl:"https://portal.azure.us",publishingProfileUrl:"https://manage.windowsazure.us/publishsettings/index",managementEndpointUrl:"https://management.core.usgovcloudapi.net",resourceManagerEndpointUrl:"https://management.usgovcloudapi.net",sqlManagementEndpointUrl:"https://management.core.usgovcloudapi.net:8443/",sqlServerHostnameSuffix:".database.usgovcloudapi.net",galleryEndpointUrl:"https://gallery.usgovcloudapi.net/",activeDirectoryEndpointUrl:"https://login-us.microsoftonline.com/",activeDirectoryResourceId:"https://management.core.usgovcloudapi.net/",activeDirectoryGraphResourceId:"https://graph.windows.net/",batchResourceId:"https://batch.core.usgovcloudapi.net/",activeDirectoryGraphApiVersion:"2013-04-05",storageEndpointSuffix:".core.usgovcloudapi.net",keyVaultDnsSuffix:".vault.usgovcloudapi.net",azureDataLakeStoreFileSystemEndpointSuffix:"N/A",azureDataLakeAnalyticsCatalogAndJobEndpointSuffix:"N/A",validateAuthority:!0},a.AzureGermanCloud={name:"AzureGermanCloud",portalUrl:"https://portal.microsoftazure.de/",publishingProfileUrl:"https://manage.microsoftazure.de/publishsettings/index",managementEndpointUrl:"https://management.core.cloudapi.de",resourceManagerEndpointUrl:"https://management.microsoftazure.de",sqlManagementEndpointUrl:"https://management.core.cloudapi.de:8443/",sqlServerHostnameSuffix:".database.cloudapi.de",galleryEndpointUrl:"https://gallery.cloudapi.de/",activeDirectoryEndpointUrl:"https://login.microsoftonline.de/",activeDirectoryResourceId:"https://management.core.cloudapi.de/",activeDirectoryGraphResourceId:"https://graph.cloudapi.de/",batchResourceId:"https://batch.microsoftazure.de/",activeDirectoryGraphApiVersion:"2013-04-05",storageEndpointSuffix:".core.cloudapi.de",keyVaultDnsSuffix:".vault.microsoftazure.de",azureDataLakeStoreFileSystemEndpointSuffix:"N/A",azureDataLakeAnalyticsCatalogAndJobEndpointSuffix:"N/A",validateAuthority:!0},t.AzureEnvironment=a}]);
//# sourceMappingURL=azureEnvironmentBundle.js.map

typings/lib/azureEnvironment.d.ts
export interface AzureEnvironmentParameters {
    /**
     * The Environment name.
     */
    readonly name: string;
    /**
     * The management portal URL.
     */
    readonly portalUrl: string;
    /**
     * The management service endpoint.
     */
    readonly managementEndpointUrl: string;
    /**
     * The resource management endpoint.
     */
    readonly resourceManagerEndpointUrl: string;
    /**
     * The Active Directory login endpoint.
     */
    readonly activeDirectoryEndpointUrl: string;
    /**
     * The resource ID to obtain AD tokens for (token audience).
     */
    readonly activeDirectoryResourceId: string;
    /**
     * The publish settings file URL.
     */
    readonly publishingProfileUrl?: string;
    /**
     * The sql server management endpoint for mobile commands.
     */
    readonly sqlManagementEndpointUrl?: string;
    /**
     * The dns suffix for sql servers.
     */
    readonly sqlServerHostnameSuffix?: string;
    /**
     * The template gallery endpoint.
     */
    readonly galleryEndpointUrl?: string;
    /**
     * The Active Directory resource ID.
     */
    readonly activeDirectoryGraphResourceId?: string;
    /**
     * The batch resource ID.
     */
    readonly batchResourceId?: string;
    /**
     * The Active Directory api version.
     */
    readonly activeDirectoryGraphApiVersion?: string;
    /**
     * The endpoint suffix for storage accounts.
     */
    readonly storageEndpointSuffix?: string;
    /**
     * The keyvault service dns suffix.
     */
    readonly keyVaultDnsSuffix?: string;
    /**
     * The data lake store filesystem service dns suffix.
     */
    readonly azureDataLakeStoreFileSystemEndpointSuffix?: string;
    /**
     * The data lake analytics job and catalog service dns suffix.
     */
    readonly azureDataLakeAnalyticsCatalogAndJobEndpointSuffix?: string;
    /**
     * Determines whether the authentication endpoint should be validated with Azure AD. Default value is true.
     */
    readonly validateAuthority?: boolean;
}
export declare class AzureEnvironment {
    /**
     * The Environment name.
     */
    readonly name: string;
    /**
     * The management portal URL.
     */
    readonly portalUrl: string;
    /**
     * The management service endpoint.
     */
    readonly managementEndpointUrl: string;
    /**
     * The resource management endpoint.
     */
    readonly resourceManagerEndpointUrl: string;
    /**
     * The Active Directory login endpoint.
     */
    readonly activeDirectoryEndpointUrl: string;
    /**
     * The resource ID to obtain AD tokens for (token audience).
     */
    readonly activeDirectoryResourceId: string;
    /**
     * The publish settings file URL.
     */
    readonly publishingProfileUrl?: string;
    /**
     * The sql server management endpoint for mobile commands.
     */
    readonly sqlManagementEndpointUrl?: string;
    /**
     * The dns suffix for sql servers.
     */
    readonly sqlServerHostnameSuffix?: string;
    /**
     * The template gallery endpoint.
     */
    readonly galleryEndpointUrl?: string;
    /**
     * The batch resource ID.
     */
    readonly batchResourceId?: string;
    /**
     * The Active Directory resource ID.
     */
    readonly activeDirectoryGraphResourceId?: string;
    /**
     * The Active Directory api version.
     */
    readonly activeDirectoryGraphApiVersion?: string;
    /**
     * The endpoint suffix for storage accounts.
     */
    readonly storageEndpointSuffix?: string;
    /**
     * The keyvault service dns suffix.
     */
    readonly keyVaultDnsSuffix?: string;
    /**
     * The data lake store filesystem service dns suffix.
     */
    readonly azureDataLakeStoreFileSystemEndpointSuffix?: string;
    /**
     * The data lake analytics job and catalog service dns suffix.
     */
    readonly azureDataLakeAnalyticsCatalogAndJobEndpointSuffix?: string;
    /**
     * Determines whether the authentication endpoint should be validated with Azure AD. Default value is true.
     */
    readonly validateAuthority: boolean;
    constructor(parameters: AzureEnvironmentParameters);
    static add(parameters: AzureEnvironmentParameters): void;
    static get(name: string): AzureEnvironment;
    static readonly Azure: {
        name: string;
        portalUrl: string;
        publishingProfileUrl: string;
        managementEndpointUrl: string;
        resourceManagerEndpointUrl: string;
        sqlManagementEndpointUrl: string;
        sqlServerHostnameSuffix: string;
        galleryEndpointUrl: string;
        activeDirectoryEndpointUrl: string;
        activeDirectoryResourceId: string;
        activeDirectoryGraphResourceId: string;
        batchResourceId: string;
        activeDirectoryGraphApiVersion: string;
        storageEndpointSuffix: string;
        keyVaultDnsSuffix: string;
        azureDataLakeStoreFileSystemEndpointSuffix: string;
        azureDataLakeAnalyticsCatalogAndJobEndpointSuffix: string;
        validateAuthority: boolean;
    };
    static readonly AzureChina: {
        name: string;
        portalUrl: string;
        publishingProfileUrl: string;
        managementEndpointUrl: string;
        resourceManagerEndpointUrl: string;
        sqlManagementEndpointUrl: string;
        sqlServerHostnameSuffix: string;
        galleryEndpointUrl: string;
        activeDirectoryEndpointUrl: string;
        activeDirectoryResourceId: string;
        activeDirectoryGraphResourceId: string;
        activeDirectoryGraphApiVersion: string;
        batchResourceId: string;
        storageEndpointSuffix: string;
        keyVaultDnsSuffix: string;
        azureDataLakeStoreFileSystemEndpointSuffix: string;
        azureDataLakeAnalyticsCatalogAndJobEndpointSuffix: string;
        validateAuthority: boolean;
    };
    static readonly AzureUSGovernment: {
        name: string;
        portalUrl: string;
        publishingProfileUrl: string;
        managementEndpointUrl: string;
        resourceManagerEndpointUrl: string;
        sqlManagementEndpointUrl: string;
        sqlServerHostnameSuffix: string;
        galleryEndpointUrl: string;
        activeDirectoryEndpointUrl: string;
        activeDirectoryResourceId: string;
        activeDirectoryGraphResourceId: string;
        batchResourceId: string;
        activeDirectoryGraphApiVersion: string;
        storageEndpointSuffix: string;
        keyVaultDnsSuffix: string;
        azureDataLakeStoreFileSystemEndpointSuffix: string;
        azureDataLakeAnalyticsCatalogAndJobEndpointSuffix: string;
        validateAuthority: boolean;
    };
    static readonly AzureGermanCloud: {
        name: string;
        portalUrl: string;
        publishingProfileUrl: string;
        managementEndpointUrl: string;
        resourceManagerEndpointUrl: string;
        sqlManagementEndpointUrl: string;
        sqlServerHostnameSuffix: string;
        galleryEndpointUrl: string;
        activeDirectoryEndpointUrl: string;
        activeDirectoryResourceId: string;
        activeDirectoryGraphResourceId: string;
        batchResourceId: string;
        activeDirectoryGraphApiVersion: string;
        storageEndpointSuffix: string;
        keyVaultDnsSuffix: string;
        azureDataLakeStoreFileSystemEndpointSuffix: string;
        azureDataLakeAnalyticsCatalogAndJobEndpointSuffix: string;
        validateAuthority: boolean;
    };
}


lib/azureEnvironment.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

export interface AzureEnvironmentParameters {
  /**
   * The Environment name.
   */
  readonly name: string;

  /**
   * The management portal URL.
   */
  readonly portalUrl: string;

  /**
   * The management service endpoint.
   */
  readonly managementEndpointUrl: string;

  /**
   * The resource management endpoint.
   */
  readonly resourceManagerEndpointUrl: string;

  /**
   * The Active Directory login endpoint.
   */
  readonly activeDirectoryEndpointUrl: string;

  /**
   * The resource ID to obtain AD tokens for (token audience).
   */
  readonly activeDirectoryResourceId: string;

  /**
   * The publish settings file URL.
   */
  readonly publishingProfileUrl?: string;

  /**
   * The sql server management endpoint for mobile commands.
   */
  readonly sqlManagementEndpointUrl?: string;

  /**
   * The dns suffix for sql servers.
   */
  readonly sqlServerHostnameSuffix?: string;

  /**
   * The template gallery endpoint.
   */
  readonly galleryEndpointUrl?: string;

  /**
   * The Active Directory resource ID.
   */
  readonly activeDirectoryGraphResourceId?: string;

  /**
   * The batch resource ID.
   */
  readonly batchResourceId?: string;

  /**
   * The Active Directory api version.
   */
  readonly activeDirectoryGraphApiVersion?: string;

  /**
   * The endpoint suffix for storage accounts.
   */
  readonly storageEndpointSuffix?: string;

  /**
   * The keyvault service dns suffix.
   */
  readonly keyVaultDnsSuffix?: string;

  /**
   * The data lake store filesystem service dns suffix.
   */
  readonly azureDataLakeStoreFileSystemEndpointSuffix?: string;

  /**
   * The data lake analytics job and catalog service dns suffix.
   */
  readonly azureDataLakeAnalyticsCatalogAndJobEndpointSuffix?: string;

  /**
   * Determines whether the authentication endpoint should be validated with Azure AD. Default value is true.
   */
  readonly validateAuthority?: boolean;
}

export class AzureEnvironment {
  /**
   * The Environment name.
   */
  readonly name: string;

  /**
   * The management portal URL.
   */
  readonly portalUrl: string;

  /**
   * The management service endpoint.
   */
  readonly managementEndpointUrl: string;

  /**
   * The resource management endpoint.
   */
  readonly resourceManagerEndpointUrl: string;

  /**
   * The Active Directory login endpoint.
   */
  readonly activeDirectoryEndpointUrl: string;

  /**
   * The resource ID to obtain AD tokens for (token audience).
   */
  readonly activeDirectoryResourceId: string;

  /**
   * The publish settings file URL.
   */
  readonly publishingProfileUrl?: string;

  /**
   * The sql server management endpoint for mobile commands.
   */
  readonly sqlManagementEndpointUrl?: string;

  /**
   * The dns suffix for sql servers.
   */
  readonly sqlServerHostnameSuffix?: string;

  /**
   * The template gallery endpoint.
   */
  readonly galleryEndpointUrl?: string;

  /**
   * The batch resource ID.
   */
  readonly batchResourceId?: string;

  /**
   * The Active Directory resource ID.
   */
  readonly activeDirectoryGraphResourceId?: string;

  /**
   * The Active Directory api version.
   */
  readonly activeDirectoryGraphApiVersion?: string;

  /**
   * The endpoint suffix for storage accounts.
   */
  readonly storageEndpointSuffix?: string;

  /**
   * The keyvault service dns suffix.
   */
  readonly keyVaultDnsSuffix?: string;

  /**
   * The data lake store filesystem service dns suffix.
   */
  readonly azureDataLakeStoreFileSystemEndpointSuffix?: string;

  /**
   * The data lake analytics job and catalog service dns suffix.
   */
  readonly azureDataLakeAnalyticsCatalogAndJobEndpointSuffix?: string;

  /**
   * Determines whether the authentication endpoint should be validated with Azure AD. Default value is true.
   */
  readonly validateAuthority: boolean = true;

  constructor(parameters: AzureEnvironmentParameters) {

    if (!parameters || typeof parameters !== "object") {
      throw new Error("'parameters' is a required parameter and must be of type 'object'.");
    }

    // Validate required parameters
    const requiredParams = ["name", "portalUrl", "managementEndpointUrl", "resourceManagerEndpointUrl",
      "activeDirectoryEndpointUrl", "activeDirectoryResourceId"];
    requiredParams.forEach(function (param) {
      if (!(<any>parameters)[param] || typeof (<any>parameters)[param].valueOf() !== "string") {
        throw new Error(`Please provide "${param}" for the environment and it must be of type "string".`);
      }
    });

    this.name = parameters.name;
    this.portalUrl = parameters.portalUrl;
    this.managementEndpointUrl = parameters.managementEndpointUrl;
    this.resourceManagerEndpointUrl = parameters.resourceManagerEndpointUrl;
    this.activeDirectoryEndpointUrl = parameters.activeDirectoryEndpointUrl;
    this.activeDirectoryResourceId = parameters.activeDirectoryResourceId;
    if (this.activeDirectoryGraphApiVersion) {
      this.activeDirectoryGraphApiVersion = parameters.activeDirectoryGraphApiVersion;
    }
    if (this.activeDirectoryGraphResourceId) {
      this.activeDirectoryGraphResourceId = parameters.activeDirectoryGraphResourceId;
    }
    if (this.azureDataLakeAnalyticsCatalogAndJobEndpointSuffix) {
      this.azureDataLakeAnalyticsCatalogAndJobEndpointSuffix = parameters.azureDataLakeAnalyticsCatalogAndJobEndpointSuffix;
    }

    if (this.azureDataLakeStoreFileSystemEndpointSuffix) {
      this.azureDataLakeStoreFileSystemEndpointSuffix = parameters.azureDataLakeStoreFileSystemEndpointSuffix;
    }

    if (this.batchResourceId) {
      this.batchResourceId = parameters.batchResourceId;
    }

    if (this.galleryEndpointUrl) {
      this.galleryEndpointUrl = parameters.galleryEndpointUrl;
    }

    if (this.keyVaultDnsSuffix) {
      this.keyVaultDnsSuffix = parameters.keyVaultDnsSuffix;
    }

    if (this.publishingProfileUrl) {
      this.publishingProfileUrl = parameters.publishingProfileUrl;
    }

    if (this.sqlManagementEndpointUrl) {
      this.sqlManagementEndpointUrl = parameters.sqlManagementEndpointUrl;
    }

    if (this.sqlServerHostnameSuffix) {
      this.sqlServerHostnameSuffix = parameters.sqlServerHostnameSuffix;
    }

    if (this.storageEndpointSuffix) {
      this.storageEndpointSuffix = parameters.storageEndpointSuffix;
    }
  }

  static add(parameters: AzureEnvironmentParameters): void {
    const envContainer: { [name: string]: AzureEnvironment } = {};
    const envObj = new AzureEnvironment(parameters);
    envContainer[parameters.name] = envObj;
    Object.assign(AzureEnvironment, envContainer);
    return;
  }

  static get(name: string): AzureEnvironment {
    if (!name) {
      throw new TypeError("name cannot be null or undefined and must be of type string.");
    }
    return (AzureEnvironment as any)[name];
  }

  static readonly Azure = {
    name: "Azure",
    portalUrl: "https://portal.azure.com",
    publishingProfileUrl: "https://go.microsoft.com/fwlink/?LinkId=254432",
    managementEndpointUrl: "https://management.core.windows.net",
    resourceManagerEndpointUrl: "https://management.azure.com/",
    sqlManagementEndpointUrl: "https://management.core.windows.net:8443/",
    sqlServerHostnameSuffix: ".database.windows.net",
    galleryEndpointUrl: "https://gallery.azure.com/",
    activeDirectoryEndpointUrl: "https://login.microsoftonline.com/",
    activeDirectoryResourceId: "https://management.core.windows.net/",
    activeDirectoryGraphResourceId: "https://graph.windows.net/",
    batchResourceId: "https://batch.core.windows.net/",
    activeDirectoryGraphApiVersion: "2013-04-05",
    storageEndpointSuffix: ".core.windows.net",
    keyVaultDnsSuffix: ".vault.azure.net",
    azureDataLakeStoreFileSystemEndpointSuffix: "azuredatalakestore.net",
    azureDataLakeAnalyticsCatalogAndJobEndpointSuffix: "azuredatalakeanalytics.net",
    validateAuthority: true
  };
  static readonly AzureChina = {
    name: "AzureChina",
    portalUrl: "https://portal.azure.cn",
    publishingProfileUrl: "https://go.microsoft.com/fwlink/?LinkID=301774",
    managementEndpointUrl: "https://management.core.chinacloudapi.cn",
    resourceManagerEndpointUrl: "https://management.chinacloudapi.cn",
    sqlManagementEndpointUrl: "https://management.core.chinacloudapi.cn:8443/",
    sqlServerHostnameSuffix: ".database.chinacloudapi.cn",
    galleryEndpointUrl: "https://gallery.chinacloudapi.cn/",
    activeDirectoryEndpointUrl: "https://login.chinacloudapi.cn/",
    activeDirectoryResourceId: "https://management.core.chinacloudapi.cn/",
    activeDirectoryGraphResourceId: "https://graph.chinacloudapi.cn/",
    activeDirectoryGraphApiVersion: "2013-04-05",
    batchResourceId: "https://batch.chinacloudapi.cn/",
    storageEndpointSuffix: ".core.chinacloudapi.cn",
    keyVaultDnsSuffix: ".vault.azure.cn",
    // TODO: add dns suffixes for the china cloud for datalake store and datalake analytics once they are defined.
    azureDataLakeStoreFileSystemEndpointSuffix: "N/A",
    azureDataLakeAnalyticsCatalogAndJobEndpointSuffix: "N/A",
    validateAuthority: true
  };
  static readonly AzureUSGovernment = {
    name: "AzureUSGovernment",
    portalUrl: "https://portal.azure.us",
    publishingProfileUrl: "https://manage.windowsazure.us/publishsettings/index",
    managementEndpointUrl: "https://management.core.usgovcloudapi.net",
    resourceManagerEndpointUrl: "https://management.usgovcloudapi.net",
    sqlManagementEndpointUrl: "https://management.core.usgovcloudapi.net:8443/",
    sqlServerHostnameSuffix: ".database.usgovcloudapi.net",
    galleryEndpointUrl: "https://gallery.usgovcloudapi.net/",
    activeDirectoryEndpointUrl: "https://login.microsoftonline.us/",
    activeDirectoryResourceId: "https://management.core.usgovcloudapi.net/",
    activeDirectoryGraphResourceId: "https://graph.windows.net/",
    batchResourceId: "https://batch.core.usgovcloudapi.net/",
    activeDirectoryGraphApiVersion: "2013-04-05",
    storageEndpointSuffix: ".core.usgovcloudapi.net",
    keyVaultDnsSuffix: ".vault.usgovcloudapi.net",
    azureDataLakeStoreFileSystemEndpointSuffix: "N/A",
    azureDataLakeAnalyticsCatalogAndJobEndpointSuffix: "N/A",
    validateAuthority: true
  };
  static readonly AzureGermanCloud = {
    name: "AzureGermanCloud",
    portalUrl: "https://portal.microsoftazure.de/",
    publishingProfileUrl: "https://manage.microsoftazure.de/publishsettings/index",
    managementEndpointUrl: "https://management.core.cloudapi.de",
    resourceManagerEndpointUrl: "https://management.microsoftazure.de",
    sqlManagementEndpointUrl: "https://management.core.cloudapi.de:8443/",
    sqlServerHostnameSuffix: ".database.cloudapi.de",
    galleryEndpointUrl: "https://gallery.cloudapi.de/",
    activeDirectoryEndpointUrl: "https://login.microsoftonline.de/",
    activeDirectoryResourceId: "https://management.core.cloudapi.de/",
    activeDirectoryGraphResourceId: "https://graph.cloudapi.de/",
    batchResourceId: "https://batch.microsoftazure.de/",
    activeDirectoryGraphApiVersion: "2013-04-05",
    storageEndpointSuffix: ".core.cloudapi.de",
    keyVaultDnsSuffix: ".vault.microsoftazure.de",
    azureDataLakeStoreFileSystemEndpointSuffix: "N/A",
    azureDataLakeAnalyticsCatalogAndJobEndpointSuffix: "N/A",
    validateAuthority: true
  };
}
